README.md
Description
===========

* Installs OSEv3 and related packages.

Requirements
============

## Platform ##

* Tested on Red Hat RHEL 7.4
* Tested on Centos 7.4

## Openshift Version ##

* Support OSE version from 3.4+
* Support Origin version from 1.4+
* Default the installation to 3.10

**Highly recommended**: 

Explicitly set `node['is_apaas_openshift_cookbook']['ose_version']`, `node['is_apaas_openshift_cookbook']['ose_major_version']`
and ideally `node['is_apaas_openshift_cookbook']['docker_version']` to be safe when a major version is released on the
CentOS PaaS repository; 

This cookbook does support upgrade between major versions (Read doc).

## New Features!

- Control upgrades between versions
- ETCD migration v2 to v3 (Mandatory before upgrading to 3.7)
- Node vars override variables
- Add or remove ETCD members from ETCD cluster
- Early support for 3.10 (Work still ongoing)

Supported Node Variables:

| NAME |
| ---------------- |
| deploy_dnsmasq |
| custom_origin-dns |
| custom_origin_location |
| osn_cluster_dns_ip |

In the following example, DNSMASQ will only be deployed on mynodeoverride.domain.local (Being overrided by local vars)

```json
  "override_attributes": {
    "is_apaas_openshift_cookbook": {
      "deploy_dnsmasq": false,
      "node_servers": [
        {
          "fqdn": "mynode.domain.local",
          "ipaddress": "1.1.1.1"
        },
        {
          "fqdn": "mynodeoverride.domain.local",
          "ipaddress": "1.2.2.2",
          "deploy_dnsmasq": true
        }
      ],
      ....
    }
  }
```

### Supported version
[x] 1.5 to 3.6

[x] 3.6 to 3.7

[x] 3.7 to 3.9 (Need to add repository for 3.8 as well (Double upgrades)

### Control Upgrade

Control upgrade requires variables to be set before running.

*The CHEF run will ignore any upgrade unless all variables are correctly set*

Variables:

| NAME | PURPOSE | Default value | Mandatory |
| ---------------- | ------------------------------- | ------------------ | ---------- |
| control_upgrade | Execute an upgrade     | `false`    | `YES` |
| control_upgrade_version | Target version (15,36,37,39)        |`""` |`YES`|
| control_upgrade_flag | Location of the control upgrade flag | `"/to_be_replaced"`  | `YES` |
| upgrade_repos | Target YUM repo | `""` | `NO` |

```json
  "override_attributes": {
    "is_apaas_openshift_cookbook": {
      "control_upgrade": true,
      "control_upgrade_version": "39",
      "control_upgrade_flag": "/tmp/ready",
      "upgrade_repos": [
        {
          "name": "centos-openshift-origin38",
          "baseurl": "http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin38/",
          "gpgcheck": false
        },
        {
          "name": "centos-openshift-origin39",
          "baseurl": "http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin39/",
          "gpgcheck": false
        }
      ]
    }
  }
```

[Read more about the upgrade process here!](docs/upgrade-process.md)

Test Matrix
===========

| Platform   | OSE 3.10 | OSE 3.9.0 | OSE 3.7.0 | OSE 3.6.1 | OSE 1.5.1 |
| --------   | -------- | --------- | --------- | --------- | --------- |
| centos 7.4 "cluster native" | ONGOING | PASS   | PASS      | PASS      | PASS      |
| centos 7.4 "standalone" | Not supported | Not supported   | Not supported      | PASS      | PASS      |

Override Attributes
===================

[Read more about overriding attributes here!](docs/attribute-cookbook.md)

#### Structure ####
-------------------

* `node['is_apaas_openshift_cookbook']['openshift_master_identity_provider']['HTPasswdPasswordIdentityProvider']`

```json
{
"name" : "htpasswd_auth", 
"login" : true, 
"challenge" : true, 
"kind" : "HTPasswdPasswordIdentityProvider", 
"filename" : "/etc/openshift/openshift-passwd"
}
```

* `node['is_apaas_openshift_cookbook']['openshift_master_identity_provider']['LDAPPasswordIdentityProvider']`

```json
{
"name" : "ldap_identity", 
"login" : true, 
"challenge" : true, 
"kind" : "LDAPPasswordIdentityProvider", 
"ldap_server" : "ldap.domain.local",
 "ldap_bind_dn" : "", 
 "ldap_bind_password" : "",
 "ldap_insecure" : true, 
 "ldap_base_ou" : "OU=people,DC=domain,DC=local", 
 "ldap_preferred_username" : "uid"
}
```

* `node['is_apaas_openshift_cookbook']['openshift_master_asset_config']`

```json
{
 "extensionStylesheets":["/path/to/css"],
 "extensionScripts":["/path/to/script"],
 "extensions":"/path/to/my_images",
 "templates":"/path/to/template"
}
```

* `node['is_apaas_openshift_cookbook']['openshift_node_docker-storage']`

```json
{
 "DEVS": ["/dev/sdb","/dev/sdc"],
 "VG": "docker-vg",
 "DATA_SIZE": "80%FREE"
}
```

* `node['is_apaas_openshift_cookbook']['docker_log_driver']`

Set to `'json-file'` (default), `'journald'` or any other supported [docker log driver](https://docs.docker.com/engine/admin/logging/overview/).
Set to '' to disable it.

* `node['is_apaas_openshift_cookbook']['docker_log_options']`

Assuming `node['is_apaas_openshift_cookbook']['docker_log_driver']` is `'json-file'` (the default):

```json
{
 "max-size": "50M",
 "max-file": "3"
}
```

Any option can be set, as long as they are supported by the current [docker log driver](https://docs.docker.com/engine/admin/logging/overview/).


* `node['is_apaas_openshift_cookbook']['openshift_master_named_certificates']`

* CN or SAN names are automatically detected from the certificate file.

```json
[
  {
   "certfile": "/etc/where_is/my_certfile",
   "keyfile": "/etc/where_is/my_keyfile"
  }
]
```

* `node['is_apaas_openshift_cookbook']['persistent_storage']`

* They key called 'claim' is optional and will automatically create a PersistentVolumeClaim within a specified namespace.
 
```json
{
  "name": "registry",
  "capacity": "5Gi",
  "access_modes": "ReadWriteMany",
  "server": "core.domain.local",
  "path": "/var/nfs/registry",
  "claim": {
    "namespace": "default"
  }
}
```

* `node['is_apaas_openshift_cookbook']['openshift_hosted_cluster_metrics']`

Any option can be set, as long as they are supported by the current [Metrics deployer template](https://docs.openshift.com/container-platform/latest/install_config/cluster_metrics.html#metrics-ansible-variables).

Full list of attributes can be found [here](https://raw.githubusercontent.com/IshentRas/is_apaas_openshift_cookbook/master/attributes/metrics.rb).

We only support 1 cassandra POD.
Once the metrics addon is deployed, it creates a flag (openshift_hosted_metrics_flag) on the first master to avoid having CHEF re-creating the resources over and over.

Example of options for deploying metrics:
```json
{
  "openshift_hosted_cluster_metrics": true,
  "openshift_metrics_heapster_requests_memory": "1Gi",
  "openshift_metrics_image_version": "v3.9",
  "openshift_metrics_cassandra_storage_type": "dynamic",
  "openshift_metrics_hawkular_user_write_access": false,
  "....."
}
```

Example of removing metrics components:
```json
{
  "openshift_hosted_cluster_metrics": true,
  "openshift_metrics_install_metrics": false,
  "....."
}
```

* `node['is_apaas_openshift_cookbook']['openshift_hosted_cluster_logging']`

Any option can be set, as long as they are supported by the current [Logging deployer template](https://docs.openshift.com/container-platform/latest/install_config/aggregate_logging.html#aggregate-logging-ansible-variables).

Full list of attributes can be found [here](https://raw.githubusercontent.com/IshentRas/is_apaas_openshift_cookbook/master/attributes/logging.rb).

We only support 1 elasticsearch POD.
Once the logging addon is deployed, it creates a flag (openshift_hosted_logging_flag) on the first master to avoid having CHEF re-creating the resources over and over.

Example of options for deploying logging:
```json
{
  "openshift_hosted_cluster_logging": true,
  "openshift_logging_es_memory_limit": "1Gi",
  "openshift_logging_kibana_memory_limit": "512Mi",
  "openshift_logging_image_version": "v3.9",
  "....."
}
```

Example of removing metrics components:
```json
{
  "openshift_hosted_cluster_logging": true,
  "openshift_logging_install_logging": false,
  "....."
}
```

Example of overriding the default kubelet options:
* `node['is_apaas_openshift_cookbook']['openshift_node_kubelet_args_custom']`
Any option can be set, as long as they are supported by current [Kubelet Options](https://kubernetes.io/docs/admin/kubelet/).

```json
{
  "openshift_node_kubelet_args_custom": {
    "pods-per-core": "5",
    "image-gc-high-threshold": "85",
    "resolv-conf": "/etc/resolv.conf"
  },
  "....."
}
```

Example of enabling a specific "feature-gate": enabling the [persistent local volumes feature](https://docs.openshift.com/container-platform/3.7/install_config/configuring_local.html#local-volume-enabling-local-volumes) (OSE v3.7 and up):
```json
{
  "openshift_master_api_server_args_custom": {
    "feature-gates": ["PersistentLocalVolumes=true"]
  },
  "openshift_master_controller_args_custom": {
    "feature-gates": ["PersistentLocalVolumes=true"]
  },
  "openshift_node_kubelet_args_custom": {
    "feature-gates": ["PersistentLocalVolumes=true"]
  }
}
```

* `node['is_apaas_openshift_cookbook']['openshift_hosted_router_deploy_shards']`

Any ENV options can be set for the sharding router, as long as they are supported by the current documentation [Set ENV for router sharding](https://docs.openshift.com/container-platform/latest/install_config/router/default_haproxy_router.html#modifying-router-shards).

** The dedicated service account must have access to the hostnetwork SCC **

** custom_router (Deploying custom template) is optional and should be omitted if not wanted

Example of options for deploying router sharding:
```json
{
  "....."
  "openshift_hosted_router_deploy_shards": true,
  "openshift_hosted_router_shard":[
    {
      "namespace": "custom-shard1",
      "service_account": "shard1",
      "selector": "region=shard1",
      "env": [
        "ROUTE_LABELS='shard1=yes'"
      ]
    },
    {
      "namespace": "custom-shard2",
      "service_account": "shard2",
      "selector": "region=shard2",
      "custom_router": true,
      "custom_router_file": "/folder/custom_template_forshard2",
      "env": [
        "NAMESPACE_LABELS='dept != finance'"
      ]
    }
  ],
  "openshift_common_service_accounts_additional": [
    { 
      "name": "shard1", 
      "namespace": "shard1", 
      "scc": ["hostnetwork"] 
    },
    { 
      "name": "shard2", 
      "namespace": "custom-shard2", 
      "scc": ["hostnetwork"] 
    }
  ],
  "....."
}
```

Example of overriding the Audit Configuration Parameters for the `openshift_master_auditconfig` Hash

| Key Name   | Description |
| --------   | --------- |
| enable | A boolean to enable or disable audit logs. Default is `false`. |
| audit-file | File path where the requests should be logged to. If not set, logs are printed to master logs. |
| max-retention-day | Specifies maximum number of days to retain old audit log files based on the time stamp encoded in their filename. |
| max-file-number | Specifies the maximum number of old audit log files to retain. |
| max-file-size | Specifies maximum size in megabytes of the log file before it gets rotated. Defaults to 100MB. |

```json
{
  ...
  "openshift_master_auditconfig": {
    "enable": true,
    "audit-file": "/var/log/audit-ocp",
    "max-file-size": "50",
    "max-file-number": "10"
  },
  ...
}
```

Example of overriding the Audit Configuration Parameters for the `openshift_master_auditconfig` Hash

| Key Name   | Description |
| --------   | --------- |
| enable | A boolean to enable or disable audit logs. Default is `false`. |
| audit-file | File path where the requests should be logged to. If not set, logs are printed to master logs. |
| max-retention-day | Specifies maximum number of days to retain old audit log files based on the time stamp encoded in their filename. |
| max-file-number | Specifies the maximum number of old audit log files to retain. |
| max-file-size | Specifies maximum size in megabytes of the log file before it gets rotated. Defaults to 100MB. |

```json
{
  ...
  "openshift_master_auditconfig": {
    "enable": true,
    "audit-file": "/var/log/audit-ocp",
    "max-file-size": "50",
    "max-file-number": "10"
  },
  ...
}
```

Example of overriding the setting for global builds:
Any option can be set, as long as they are supported by current [Global Build Defaults and Overrides](https://docs.openshift.com/container-platform/latest/install_config/build_defaults_overrides.html).

**When setting CPU/Memory via `openshift_builddefaults_resources_*`, `openshift_builddefaults_resources` needs to be set to `true`**

```json
{
  "....."
  "openshift_builddefaults_git_http_proxy": "http://USER:PASSWORD@HOST:PORT",
  "openshift_builddefaults_git_https_proxy": "https://USER:PASSWORD@HOST:PORT",
  "openshift_builddefaults_git_no_proxy": "mycorp.com",
  "openshift_builddefaults_env": [
    {
      "http_proxy": "http://proxy.example.com.redhat.com:3128"
    },
    {
      "NO_PROXY": "ose3-master.example.com"
    }
  ],
  "openshift_builddefaults_image_labels": [
    {
      "name": "imagelabelname1",
      "value": "imagelabelvalue1"
    }
  ],
  "openshift_builddefaults_nodeselectors": {
    "nodelabel1": "nodelabelvalue1",
    "nodelabel2": "nodelabelvalue2"
  },
  "openshift_builddefaults_annotations": {
    "annotationkey1": "annotationvalue1"
  },
  "openshift_buildoverrides_force_pull": "true",
  "openshift_buildoverrides_image_labels": [
    {
      "name": "imagelabelname1",

CHANGELOG.md
# Openshift 3 Cookbook CHANGELOG
This file is used to list changes made in each version of the Openshift 3 cookbook.

## v2.1.31
### Bug
- Make sure the services (api, controller and node) restart when performing an asynchronous update

### Improvement
- Add more ec2 userdata defined attributes (openshift_hosted_router_certfile openshift_hosted_router_keyfile openshift_master_named_certificates)

## v2.1.27-29
### Bug
- Enable service discovery when uninstalling node

### Improvement
- Add capability for defining its additional plugin configutations within admissionConfig (https://docs.openshift.com/container-platform/latest/architecture/additional_concepts/admission_controllers.html#admission-controllers-customizable-admission-plug-ins)

### TechPreview (WIP)
- Add capability to read variables from userdata at EC2 level

## v2.1.24-26
### Improvement
- Give capability to trigger uninstall of nodes via flag file. (Action is ignored if node part of Control Plane)

## v2.1.23
### Bug
- Ensure ETCD tag is added when using containerised deployment

## v2.1.22
### Bug
- Fix issue when performing ETCD migration v2 to v3

## v2.1.20-21
### Bug
- Make sure docker staorage is reset before first startup (Left over previous installations)

### Improvement
- Capacity to define labels against EC2 instances via user-data (openshift_node_user_data and ocp_labels)
- Add openshift_buildoverrides_tolerations capability 3.9+
- Add openshift-infra-selector variable to enforce default node selector for openshift-infra project (Default to region=infra. Only if set_openshift-infra_selector is set to true)
- Give the possibilty to use a search filter when using LDAPPasswordIdentityProvider

## v2.1.17-19
### Bug
- Make sure running actions from 1st master are done using loopback address (Avoiding dropping connectivity from LB)
- Make sure master servers use loopback address for their Api Config
- Add missing CheckVolumeBinding predicate (fix Local Persistent Volumes for OSE 3.9)

### Improvement
- Simplify logic around discovering first master and first etcd when using role discovery
- Add yum options for core packages
- Give the possibility to override global variables at the node level for upgrade_docker_version and docker_version

## v2.1.10-16
### Bug
- Fix issue when upgrading cluster to 3.6
- Improve logic for migrating ETCD v2 to v3 schema

## v2.1.9
### Improvement
- Create backup directory against /var/lib/etcd for ETCD v2-v3 migration
- General improvement for the controlled planed upgrade mechanism
 
## v2.1.8
### Improvement
- Add more time for ETCD migration process
- Give the possibility of deleting the ETCD recovery directory on certificate server (adhoc_clean_etcd_flag)
- Give the possibility to skip the oc adm migration storage phase pre/post upgrade (skip_migration_storage)
- Give the possibility to customise the name of the resources for oc adm migration storage (customised_storage & customised_resources)

### Bug
- Use IP addresses for checking ETCD health whilst recovering the cluster

## v2.1.7
### Improvement
- Add Time parameters for ETCD around heartbeat interval (etcd_heartbeat_interval) and election timeout (etcd_election_timeout)

## v2.1.6
### Improvement
- Add safeguards for ETCD recovery process
- Enforce docker restart process when changing OCP CA certificate

## v2.1.4-5
### Bug
- Make sure adding ETCD server(s) does not skip non-registered ETCD
- Make sure ETCD file is 644 (Needed when umask is not 022)
- Fix issue when upgrading from 3.6 to 3.7 (https://github.com/IshentRas/is_apaas_openshift_cookbook/issues/292)

## v2.1.3
### Bug
- Filter out tuned-profiles-origin-node for OSE v3.9
- Misc fixes to ng v310 recipes

## v2.1.1-2
### Improvement
- Add possibility to configure any option supported by docker-storage-setup 
- Improve installation pkgs for master and node so as to avoid unwanted automatic upgrades whilst enforcing versions

## v2.1.0
### Improvement
- Add possibility to opt-in opt-out when using sharding routers for customised template via "custom_router": true
  This key is optional and can be omitted

## v2.0.99
### Bug
- Make sure the enc file for node servers are 644
- Make sure we unmount image and container layers on-disk before resetting docker storage

## v2.0.96-98
### Improvement
- Reinstall docker whilst resetting nodes

## v2.0.91
### Improvement
- Give capability for recovering ETCD cluster
- ETCD Snapshot tuning (etcd_snapshot_count)

## v2.0.90
### Bug
- Fix logic when checking API availability
- Fix issue when disabling firewalld service for the 1st time

## v2.0.88
### Improvement
- Add logic for waiting up to 15 minutes when first time creating etcd server

## v2.0.86
### Improvement
- (FIX) Give capability for specifing personalised Admission Plug-in (openshift_master_admission_plugin_config)
- Patch any shard routers during upgrade

## v2.0.85
### Improvement
- Give capability for specifing personalised Admission Plug-in (openshift_master_admission_plugin_config)

## v2.0.83
### Bug
- Fix MISC etcd migration
- Fix incorrect openshift_master_loopback_api_url (was not proxy safe)

### Improvement
- Make SWAP disabling on nodes optional (but enabled by default).

## v2.0.82
### Bug
- Restrict permission for policy and scheduler (https://github.com/IshentRas/is_apaas_openshift_cookbook/issues/257)
- Add version control for etcd in certificate server

## v2.0.81
### Improvement
- Restrict cookbook run only to control plane and node servers

## v2.0.78
### Bug
- CRT and KEY files permissions were wrong (https://github.com/IshentRas/is_apaas_openshift_cookbook/issues/253 //github.com/IshentRas/is_apaas_openshift_cookbook/issues/254)
- Adjust the permission for OCP node main directory and ETCD

### Improvement
- Update metrics and logging to 3.9

## v2.0.77
### Bug
- CRT and KEY files permissions were wrong (https://github.com/IshentRas/is_apaas_openshift_cookbook/issues/252)
- Adjust the permission for OCP main directory

### Improvement

recipes/node.rb
#
# Cookbook Name:: is_apaas_openshift_cookbook
# Recipe:: node
#
# Copyright (c) 2015 The Authors, All Rights Reserved.

server_info = helper = OpenShiftHelper::NodeHelper.new(node)
certificate_server = server_info.certificate_server
is_node_server = server_info.on_node_server?
docker_version = node['is_apaas_openshift_cookbook']['openshift_docker_image_version']
pkg_node_to_install = node['is_apaas_openshift_cookbook']['pkg_node']

ose_major_version = node['is_apaas_openshift_cookbook']['deploy_containerized'] == true ? node['is_apaas_openshift_cookbook']['openshift_docker_image_version'] : node['is_apaas_openshift_cookbook']['ose_major_version']
path_certificate = node['is_apaas_openshift_cookbook']['use_wildcard_nodes'] ? 'wildcard_nodes.tgz.enc' : "#{node['fqdn']}.tgz.enc"

if node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_name'] && node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_item_name']
  secret_file = node['is_apaas_openshift_cookbook']['encrypted_file_password']['secret_file'] || nil
  encrypted_file_password = data_bag_item(node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_name'], node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_item_name'], secret_file)
else
  encrypted_file_password = node['is_apaas_openshift_cookbook']['encrypted_file_password']['default']
end

if is_node_server
  ruby_block 'Turn off SWAP for nodes' do
    block do
      server_info.turn_off_swap
    end
    not_if { ::File.readlines('/etc/fstab').grep(/(^[^#].*swap.*)\n/).none? }
    only_if { node['is_apaas_openshift_cookbook']['openshift_node_disable_swap_on_host'] }
  end

  file '/usr/local/etc/.firewall_node_additional.txt' do
    content node['is_apaas_openshift_cookbook']['enabled_firewall_additional_rules_node'].join("\n")
    owner 'root'
    group 'root'
  end

  node['is_apaas_openshift_cookbook']['enabled_firewall_rules_node'].each do |rule|
    iptables_rule rule do
      action :enable
      notifies :restart, 'service[iptables]', :immediately
    end
  end

  directory node['is_apaas_openshift_cookbook']['openshift_node_config_dir'] do
    recursive true
  end

  if node['is_apaas_openshift_cookbook']['deploy_containerized']
    docker_image node['is_apaas_openshift_cookbook']['openshift_docker_node_image'] do
      tag docker_version
      action :pull_if_missing
    end

    docker_image node['is_apaas_openshift_cookbook']['openshift_docker_ovs_image'] do
      tag docker_version
      action :pull_if_missing
    end

    template '/etc/systemd/system/atomic-openshift-node-dep.service' do
      source 'service_node-deps-containerized.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
    end

    template '/etc/systemd/system/atomic-openshift-node.service' do
      source 'service_node-containerized.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
      variables(ose_major_version: ose_major_version)
    end

    template '/etc/systemd/system/openvswitch.service' do
      source 'service_openvswitch-containerized.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
    end

    template '/etc/sysconfig/openvswitch' do
      source 'service_openvswitch.sysconfig.erb'
      notifies :restart, 'service[openvswitch]', :immediately unless node['is_apaas_openshift_cookbook']['upgrade']
    end
  else
    template '/etc/systemd/system/atomic-openshift-node.service' do
      source 'service_node.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
      variables(ose_major_version: ose_major_version)
      only_if { ose_major_version.split('.')[1].to_i >= 6 }
    end
  end

  sysconfig_vars = {}

  if node['is_apaas_openshift_cookbook']['openshift_cloud_provider'] == 'aws'
    if node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_name'] && node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_item_name']
      secret_file = node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['secret_file'] || nil
      aws_vars = data_bag_item(node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_name'], node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_item_name'], secret_file)

      sysconfig_vars['aws_access_key_id'] = aws_vars['access_key_id']
      sysconfig_vars['aws_secret_access_key'] = aws_vars['secret_access_key']
    end
  end

  template '/etc/sysconfig/atomic-openshift-node' do
    source 'service_node.sysconfig.erb'
    variables(sysconfig_vars)
    notifies :restart, 'service[Restart Node]', :immediately unless node['is_apaas_openshift_cookbook']['upgrade'] || Mixlib::ShellOut.new('systemctl is-enabled atomic-openshift-node').run_command.error?
  end

  pkg_node_array = pkg_node_to_install.reject { |x| x == 'tuned-profiles-atomic-openshift-node' && (node['is_apaas_openshift_cookbook']['ose_major_version'].split('.')[1].to_i >= 9 || node['is_apaas_openshift_cookbook']['control_upgrade_version'].to_i >= 39) }
  yum_package pkg_node_array do
    action :install
    version Array.new(pkg_node_array.size, node['is_apaas_openshift_cookbook']['ose_version']) unless node['is_apaas_openshift_cookbook']['ose_version'].nil?
    options node['is_apaas_openshift_cookbook']['openshift_yum_options'] unless node['is_apaas_openshift_cookbook']['openshift_yum_options'].nil?
    not_if { node['is_apaas_openshift_cookbook']['deploy_containerized'] }
    flush_cache [:before]
    retries 3
  end

  if node['is_apaas_openshift_cookbook']['asynchronous_upgrade']
    log 'Node services (Asynchronous Upgrade)' do
      level :info
      notifies :restart, 'service[Restart Node]', :immediately
      not_if { ::File.file?("/usr/local/share/info/.upgrade-#{node['is_apaas_openshift_cookbook']['ose_version']}") }
    end
    file "/usr/local/share/info/.upgrade-#{node['is_apaas_openshift_cookbook']['ose_version']}" do
      action :create_if_missing
    end
  end

  yum_package 'conntrack-tools' do
    action :install
    not_if { node['is_apaas_openshift_cookbook']['deploy_containerized'] }
    retries 3
  end

  if node['is_apaas_openshift_cookbook']['adhoc_redeploy_cluster_ca']
    Chef::Log.warn("The CLUSTER CA CERTS redeploy will be skipped for Node[#{node['fqdn']}]. Could not find the flag: #{node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag']}") unless ::File.file?(node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag'])

    ruby_block "Redeploy CA certs for Node server: #{node['fqdn']}" do
      block do
        helper.remove_dir("#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz*")
      end
      only_if { ::File.file?(node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag']) }
      notifies :delete, "file[#{node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag']}]", :immediately
      notifies :restart, 'service[Restart Node]', :delayed if ::File.file?(node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag'])
    end

    file node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag'] do
      action :nothing
    end
  end

  remote_file "Retrieve certificate from Master[#{certificate_server['fqdn']}]" do
    path "#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz.enc"
    source "http://#{certificate_server['ipaddress']}:#{node['is_apaas_openshift_cookbook']['httpd_xfer_port']}/node/generated-configs/#{path_certificate}"
    action :create_if_missing
    notifies :run, 'execute[Un-encrypt node certificate tgz files]', :immediately
    notifies :run, 'execute[Extract certificate to Node folder]', :immediately
    retries 60
    retry_delay 5
  end

  execute 'Un-encrypt node certificate tgz files' do
    command "openssl enc -d -aes-256-cbc -in #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz.enc -out #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz -k '#{encrypted_file_password}'"
    action :nothing
  end

  execute 'Extract certificate to Node folder' do
    command "tar xzf #{node['fqdn']}.tgz && chown -R root:root ."
    cwd node['is_apaas_openshift_cookbook']['openshift_node_config_dir']
    action :nothing
  end

  directory "Fix permissions on #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}" do
    path node['is_apaas_openshift_cookbook']['openshift_node_config_dir']
    owner 'root'
    group 'root'
    mode '0755'
  end

  file "Fix permissions on #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/ca.crt" do
    path ::File.join(node['is_apaas_openshift_cookbook']['openshift_node_config_dir'], 'ca.crt')
    owner 'root'
    group 'root'
    mode '0644'
  end

  remote_file '/etc/pki/ca-trust/source/anchors/openshift-ca.crt' do
    source "file://#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/ca.crt"
    notifies :run, 'ruby_block[Update ca trust]', :immediately
    sensitive true
  end

  # Use ruby_block for copying OpenShift CA to system CA trust
  ruby_block 'Update ca trust' do
    block do
      Mixlib::ShellOut.new('update-ca-trust').run_command
    end
    notifies :restart, 'service[docker]', :immediately
    notifies :run, 'execute[Wait for 30 seconds for docker services to come up]', :immediately
    action :nothing
  end

  execute 'Wait for 30 seconds for docker services to come up' do
    command 'sleep 30'
    action :nothing
    only_if { node['is_apaas_openshift_cookbook']['deploy_containerized'] }
    not_if { node['is_apaas_openshift_cookbook']['upgrade'] }
  end

  if helper.get_nodevar('deploy_dnsmasq')
    package 'NetworkManager' do
      retries 3
    end

    template '/etc/origin/node/node-dnsmasq.conf' do
      source 'node-dnsmasq.conf.erb'
      only_if { ose_major_version.split('.')[1].to_i >= 6 }
    end

    template '/etc/dnsmasq.d/origin-dns.conf' do
      source 'origin-dns.conf.erb'
      variables(
        ose_major_version: ose_major_version,
        openshift_node_dnsmasq_log_queries: helper.get_nodevar('openshift_node_dnsmasq_log_queries'),
        openshift_node_dnsmasq_maxcachettl: helper.get_nodevar('openshift_node_dnsmasq_maxcachettl'),
        openshift_node_dnsmasq_interface: helper.get_nodevar('openshift_node_dnsmasq_interface'),
        openshift_node_dnsmasq_bind_interface: helper.get_nodevar('openshift_node_dnsmasq_bind_interface')
      )
      notifies :restart, 'service[dnsmasq]', :immediately
    end

    if helper.get_nodevar('custom_origin-dns')
      remote_file 'Retrieve custom file for 99-origin-dns.sh' do
        path '/etc/NetworkManager/dispatcher.d/99-origin-dns.sh'
        source "file://#{helper.get_nodevar('custom_origin_location')}"
        owner 'root'
        group 'root'
        mode '0755'
        notifies :restart, 'service[NetworkManager]', :immediately
      end
    else
      # On some systems, NetworkManager does not exist, so ignore_failure.
      cookbook_file '/etc/NetworkManager/dispatcher.d/99-origin-dns.sh' do
        source '99-origin-dns.sh'
        owner 'root'
        group 'root'
        mode '0755'
        action :create
        ignore_failure true
        notifies :restart, 'service[NetworkManager]', :immediately
      end
    end

    ruby_block 'Setup dnsmasq' do
      block do
        f = Chef::Util::FileEdit.new('/etc/dnsmasq.conf')
        f.insert_line_if_no_match(%r{^conf-dir=/etc/dnsmasq.d}, 'conf-dir=/etc/dnsmasq.d')
        f.write_file
      end
    end

    # ignore_failure in case this fails/is not necessary
    service 'dnsmasq' do
      action %i[enable start]
      ignore_failure true
    end

    ruby_block 'Enforce running NM_CONTROLLED on host (>= 3.6)' do
      block do
        f = Chef::Util::FileEdit.new("/etc/sysconfig/network-scripts/ifcfg-#{node['network']['default_interface']}")
        f.search_file_delete_line(/^NM_CONTROLLED/)
        f.write_file
      end
      notifies :restart, 'service[NetworkManager]', :immediately
      only_if { ::File.exist?("/etc/sysconfig/network-scripts/ifcfg-#{node['network']['default_interface']}") }
      only_if { ose_major_version.split('.')[1].to_i >= 6 }
    end
  end

  template node['is_apaas_openshift_cookbook']['openshift_node_config_file'] do
    source 'node.yaml.erb'
    variables(
      osn_cluster_dns_ip: helper.get_nodevar('osn_cluster_dns_ip'),
      node_labels: helper.nodelabels,
      ose_major_version: ose_major_version,
      kubelet_args: node['is_apaas_openshift_cookbook']['openshift_node_kubelet_args_default'].merge(node['is_apaas_openshift_cookbook']['openshift_node_kubelet_args_custom'])
    )
    notifies :run, 'execute[daemon-reload]', :immediately
    notifies :enable, 'service[atomic-openshift-node]', :immediately
    notifies :restart, 'service[Restart Node]', :immediately
  end

  selinux_policy_boolean 'virt_use_nfs' do
    value true
  end

  execute 'Wait for API to become available before starting Node component' do
    command '[[ $(curl --silent --tlsv1.2 --max-time 2 -k ${MASTER_URL}/healthz/ready) =~ "ok" ]]'
    environment 'MASTER_URL' => node['is_apaas_openshift_cookbook']['openshift_master_api_url']
    retries 120
    retry_delay 1
    notifies :start, 'service[Restart Node]', :immediately unless node['is_apaas_openshift_cookbook']['upgrade'] && node['is_apaas_openshift_cookbook']['deploy_containerized']
    notifies :restart, 'service[Restart Node]', :immediately if node['is_apaas_openshift_cookbook']['upgrade'] && node['is_apaas_openshift_cookbook']['deploy_containerized']
  end

  ruby_block 'Adjust permissions for certificate and key files on Node servers' do
    block do
      run_context = Chef::RunContext.new(Chef::Node.new, {}, Chef::EventDispatch::Dispatcher.new)
      Dir.glob("#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/*").grep(/\.(?:key)$/).uniq.each do |key|
        file = Chef::Resource::File.new(key, run_context)
        file.owner 'root'
        file.group 'root'
        file.mode '0600'
        file.run_action(:create)
      end

      Dir.glob("#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/*").grep(/\.(?:crt)$/).uniq.each do |key|
        file = Chef::Resource::File.new(key, run_context)
        file.owner 'root'
        file.group 'root'
        file.mode '0640'
        file.run_action(:create)
      end
    end
  end
end
